File: javaclient/src/main/java/jagex2/io/Packet.java
package jagex2.io;

import deob.ObfuscatedName;
import jagex2.datastruct.DoublyLinkable;
import jagex2.datastruct.LinkList;

import java.math.BigInteger;

@ObfuscatedName("mb")
public class Packet extends DoublyLinkable {

	@ObfuscatedName("mb.p")
	public byte[] data;

	@ObfuscatedName("mb.q")
	public int pos;

	@ObfuscatedName("mb.r")
	public int bitPos;

	@ObfuscatedName("mb.s")
	public static int[] crctable = new int[256];

	@ObfuscatedName("mb.t")
	public static final int[] BITMASK = new int[] { 0, 1, 3, 7, 15, 31, 63, 127, 255, 511, 1023, 2047, 4095, 8191, 16383, 32767, 65535, 131071, 262143, 524287, 1048575, 2097151, 4194303, 8388607, 16777215, 33554431, 67108863, 134217727, 268435455, 536870911, 1073741823, Integer.MAX_VALUE, -1 };

	@ObfuscatedName("mb.u")
	public Isaac random;

	@ObfuscatedName("mb.v")
	public static int cacheMinCount;

	@ObfuscatedName("mb.w")
	public static int cacheMidCount;

	@ObfuscatedName("mb.x")
	public static int cacheMaxCount;

	@ObfuscatedName("mb.y")
	public static final LinkList cacheMin = new LinkList();

	@ObfuscatedName("mb.z")
	public static final LinkList cacheMid = new LinkList();

	@ObfuscatedName("mb.A")
	public static final LinkList cacheMax = new LinkList();

	@ObfuscatedName("mb.a(ZI)Lmb;")
	public static Packet alloc(int type) {
		synchronized (cacheMid) {
			Packet cached = null;
			if (type == 0 && cacheMinCount > 0) {
				cacheMinCount--;
				cached = (Packet) cacheMin.pop();
			} else if (type == 1 && cacheMidCount > 0) {
				cacheMidCount--;
				cached = (Packet) cacheMid.pop();
			} else if (type == 2 && cacheMaxCount > 0) {
				cacheMaxCount--;
				cached = (Packet) cacheMax.pop();
			}

			if (cached != null) {
				cached.pos = 0;
				return cached;
			}
		}

		Packet buf = new Packet();
		buf.pos = 0;
		if (type == 0) {
			buf.data = new byte[100];
		} else if (type == 1) {
			buf.data = new byte[5000];
		} else {
			buf.data = new byte[30000];
		}
		return buf;
	}

	@ObfuscatedName("mb.a(Z)V")
	public void release() {
		synchronized (cacheMid) {
			this.pos = 0;

			if (this.data.length == 100 && cacheMinCount < 1000) {
				cacheMin.push(this);
				cacheMinCount++;
			} else if (this.data.length == 5000 && cacheMidCount < 250) {
				cacheMid.push(this);
				cacheMidCount++;
			} else if (this.data.length == 30000 && cacheMaxCount < 50) {
				cacheMax.push(this);
				cacheMaxCount++;
			}
		}
	}

	public Packet() {
	}

	public Packet(byte[] src) {
		this.data = src;
		this.pos = 0;
	}

	@ObfuscatedName("mb.a(II)V")
	public void pIsaac(int op) {
		this.data[this.pos++] = (byte) (op + this.random.nextInt());
	}

	@ObfuscatedName("mb.a(I)V")
	public void p1(int n) {
		this.data[this.pos++] = (byte) n;
	}

	@ObfuscatedName("mb.b(I)V")
	public void p2(int n) {
		this.data[this.pos++] = (byte) (n >> 8);
		this.data[this.pos++] = (byte) n;
	}

	@ObfuscatedName("mb.b(II)V")
	public void ip2(int n) {
		this.data[this.pos++] = (byte) n;
		this.data[this.pos++] = (byte) (n >> 8);
	}

	@ObfuscatedName("mb.c(I)V")
	public void p3(int n) {
		this.data[this.pos++] = (byte) (n >> 16);
		this.data[this.pos++] = (byte) (n >> 8);
		this.data[this.pos++] = (byte) n;
	}

	@ObfuscatedName("mb.d(I)V")
	public void p4(int n) {
		this.data[this.pos++] = (byte) (n >> 24);
		this.data[this.pos++] = (byte) (n >> 16);
		this.data[this.pos++] = (byte) (n >> 8);
		this.data[this.pos++] = (byte) n;
	}

	@ObfuscatedName("mb.a(IB)V")
	public void ip4(int n) {
		this.data[this.pos++] = (byte) n;
		this.data[this.pos++] = (byte) (n >> 8);
		this.data[this.pos++] = (byte) (n >> 16);
		this.data[this.pos++] = (byte) (n >> 24);
	}

	@ObfuscatedName("mb.a(IJ)V")
	public void p8(long n) {
		this.data[this.pos++] = (byte) (n >> 56);
		this.data[this.pos++] = (byte) (n >> 48);
		this.data[this.pos++] = (byte) (n >> 40);
		this.data[this.pos++] = (byte) (n >> 32);
		this.data[this.pos++] = (byte) (n >> 24);
		this.data[this.pos++] = (byte) (n >> 16);
		this.data[this.pos++] = (byte) (n >> 8);
		this.data[this.pos++] = (byte) n;
	}

	@ObfuscatedName("mb.a(Ljava/lang/String;)V")
	public void pjstr(String text) {
		text.getBytes(0, text.length(), this.data, this.pos);
		this.pos += text.length();
		this.data[this.pos++] = 10;
	}

	@ObfuscatedName("mb.a([BIII)V")
	public void pdata(byte[] src, int len, int off) {
		for (int i = off; i < off + len; i++) {
			this.data[this.pos++] = src[i];
		}
	}

	@ObfuscatedName("mb.a(IZ)V")
	public void psize1(int size) {
		this.data[this.pos - size - 1] = (byte) size;
	}

	@ObfuscatedName("mb.c()I")
	public int g1() {
		return this.data[this.pos++] & 0xFF;
	}

	@ObfuscatedName("mb.d()B")
	public byte g1b() {
		return this.data[this.pos++];
	}

	@ObfuscatedName("mb.e()I")
	public int g2() {
		this.pos += 2;
		return ((this.data[this.pos - 2] & 0xFF) << 8) + (this.data[this.pos - 1] & 0xFF);
	}

	@ObfuscatedName("mb.f()I")
	public int g2b() {
		this.pos += 2;
		int var1 = ((this.data[this.pos - 2] & 0xFF) << 8) + (this.data[this.pos - 1] & 0xFF);
		if (var1 > 32767) {
			var1 -= 65536;
		}
		return var1;
	}

	@ObfuscatedName("mb.g()I")
	public int g3() {
		this.pos += 3;
		return ((this.data[this.pos - 3] & 0xFF) << 16) + ((this.data[this.pos - 2] & 0xFF) << 8) + (this.data[this.pos - 1] & 0xFF);
	}

	@ObfuscatedName("mb.h()I")
	public int g4() {
		this.pos += 4;
		return ((this.data[this.pos - 4] & 0xFF) << 24) + ((this.data[this.pos - 3] & 0xFF) << 16) + ((this.data[this.pos - 2] & 0xFF) << 8) + (this.data[this.pos - 1] & 0xFF);
	}

	@ObfuscatedName("mb.b(Z)J")
	public long g8() {
		long var2 = (long) this.g4() & 0xFFFFFFFFL;
		long var4 = (long) this.g4() & 0xFFFFFFFFL;
		return (var2 << 32) + var4;
	}

	@ObfuscatedName("mb.i()Ljava/lang/String;")
	public String gjstr() {
		int var1 = this.pos;
		while (this.data[this.pos++] != 10) {
		}
		return new String(this.data, var1, this.pos - var1 - 1);
	}

	@ObfuscatedName("mb.e(I)[B")
	public byte[] gjstrraw() {
		int var2 = this.pos;
		while (this.data[this.pos++] != 10) {
		}
		byte[] var3 = new byte[this.pos - var2 - 1];
		for (int var4 = var2; var4 < this.pos - 1; var4++) {
			var3[var4 - var2] = this.data[var4];
		}
		return var3;
	}

	@ObfuscatedName("mb.a(II[BI)V")
	public void gdata(int arg0, byte[] arg2, int arg3) {
		for (int var5 = arg3; var5 < arg3 + arg0; var5++) {
			arg2[var5] = this.data[this.pos++];
		}
	}

	@ObfuscatedName("mb.f(I)V")
	public void bits() {
		this.bitPos = this.pos * 8;
	}

	@ObfuscatedName("mb.c(II)I")
	public int gBit(int arg0) {
		int var3 = this.bitPos >> 3;
		int var4 = 8 - (this.bitPos & 0x7);
		int var5 = 0;
		this.bitPos += arg0;
		while (arg0 > var4) {
			var5 += (this.data[var3++] & BITMASK[var4]) << arg0 - var4;
			arg0 -= var4;
			var4 = 8;
		}
		int var7;
		if (arg0 == var4) {
			var7 = var5 + (this.data[var3] & BITMASK[var4]);
		} else {
			var7 = var5 + (this.data[var3] >> var4 - arg0 & BITMASK[arg0]);
		}
		return var7;
	}

	@ObfuscatedName("mb.g(I)V")
	public void bytes() {
		this.pos = (this.bitPos + 7) / 8;
	}

	@ObfuscatedName("mb.j()I")
	public int gsmart() {
		int var1 = this.data[this.pos] & 0xFF;
		return var1 < 128 ? this.g1() - 64 : this.g2() - 49152;
	}

	@ObfuscatedName("mb.k()I")
	public int gsmarts() {
		int var1 = this.data[this.pos] & 0xFF;
		return var1 < 128 ? this.g1() : this.g2() - 32768;
	}

	@ObfuscatedName("mb.a(Ljava/math/BigInteger;Ljava/math/BigInteger;I)V")
	public void rsaenc(BigInteger arg0, BigInteger arg1) {
		int var4 = this.pos;
		this.pos = 0;
		byte[] var5 = new byte[var4];
		this.gdata(var4, var5, 0);
		BigInteger var6 = new BigInteger(var5);
		BigInteger var7 = var6.modPow(arg0, arg1);
		byte[] var8 = var7.toByteArray();
		this.pos = 0;
		this.p1(var8.length);
		this.pdata(var8, var8.length, 0);
	}

	static {
		for (int var0 = 0; var0 < 256; var0++) {
			int var1 = var0;
			for (int var2 = 0; var2 < 8; var2++) {
				if ((var1 & 0x1) == 1) {
					var1 = var1 >>> 1 ^ 0xEDB88320;
				} else {
					var1 >>>= 0x1;
				}
			}
			crctable[var0] = var1;
		}
	}
}

File: server/webclient/src/io/Packet.ts
import DoublyLinkable from '#/datastruct/DoublyLinkable.js';
import LinkList from '#/datastruct/LinkList.js';

import Isaac from '#/io/Isaac.js';

import { bigIntModPow, bigIntToBytes, bytesToBigInt } from '#/util/JsUtil.js';

export default class Packet extends DoublyLinkable {
    private static readonly CRC32_POLYNOMIAL: number = 0xedb88320;

    private static readonly crctable: Int32Array = new Int32Array(256);
    private static readonly bitmask: Uint32Array = new Uint32Array(33);

    private static readonly cacheMin: LinkList = new LinkList();
    private static readonly cacheMid: LinkList = new LinkList();
    private static readonly cacheMax: LinkList = new LinkList();

    private static cacheMinCount: number = 0;
    private static cacheMidCount: number = 0;
    private static cacheMaxCount: number = 0;

    static {
        for (let i: number = 0; i < 32; i++) {
            Packet.bitmask[i] = (1 << i) - 1;
        }
        Packet.bitmask[32] = 0xffffffff;

        for (let i: number = 0; i < 256; i++) {
            let remainder: number = i;

            for (let bit: number = 0; bit < 8; bit++) {
                if ((remainder & 1) === 1) {
                    remainder = (remainder >>> 1) ^ Packet.CRC32_POLYNOMIAL;
                } else {
                    remainder >>>= 1;
                }
            }

            Packet.crctable[i] = remainder;
        }
    }

    static getcrc(src: Uint8Array, offset: number, length: number): number {
        let crc = 0xffffffff;
        for (let i = offset; i < length; i++) {
            crc = (crc >>> 8) ^ this.crctable[(crc ^ src[i]) & 0xff];
        }
        return ~crc;
    }

    static checkcrc(src: Uint8Array, offset: number, length: number, expected: number = 0): boolean {
        return Packet.getcrc(src, offset, length) == expected;
    }

    // constructor
    private readonly view: DataView;
    readonly data: Uint8Array;

    // runtime
    pos: number = 0;
    bitPos: number = 0;
    random: Isaac | null = null;

    constructor(src: Uint8Array | Int8Array | null) {
        if (!src) {
            throw new Error();
        }

        super();

        if (src instanceof Int8Array) {
            this.data = new Uint8Array(src);
        } else {
            this.data = src;
        }

        this.view = new DataView(this.data.buffer, this.data.byteOffset, this.data.byteLength);
    }

    get length(): number {
        return this.view.byteLength;
    }

    get available(): number {
        return this.view.byteLength - this.pos;
    }

    static alloc(type: number): Packet {
        let cached: Packet | null = null;
        if (type === 0 && Packet.cacheMinCount > 0) {
            Packet.cacheMinCount--;
            cached = Packet.cacheMin.pop() as Packet | null;
        } else if (type === 1 && Packet.cacheMidCount > 0) {
            Packet.cacheMidCount--;
            cached = Packet.cacheMid.pop() as Packet | null;
        } else if (type === 2 && Packet.cacheMaxCount > 0) {
            Packet.cacheMaxCount--;
            cached = Packet.cacheMax.pop() as Packet | null;
        }

        if (cached) {
            cached.pos = 0;
            return cached;
        }

        if (type === 0) {
            return new Packet(new Uint8Array(100));
        } else if (type === 1) {
            return new Packet(new Uint8Array(5000));
        } else {
            return new Packet(new Uint8Array(30000));
        }
    }

    release(): void {
        this.pos = 0;

        if (this.length === 100 && Packet.cacheMinCount < 1000) {
            Packet.cacheMin.push(this);
            Packet.cacheMinCount++;
        } else if (this.length === 5000 && Packet.cacheMidCount < 250) {
            Packet.cacheMid.push(this);
            Packet.cacheMidCount++;
        } else if (this.length === 30000 && Packet.cacheMaxCount < 50) {
            Packet.cacheMax.push(this);
            Packet.cacheMaxCount++;
        }
    }

    g1(): number {
        return this.view.getUint8(this.pos++);
    }

    // signed
    g1b(): number {
        return this.view.getInt8(this.pos++);
    }

    g2(): number {
        const result: number = this.view.getUint16(this.pos);
        this.pos += 2;
        return result;
    }

    // signed
    g2b(): number {
        const result: number = this.view.getInt16(this.pos);
        this.pos += 2;
        return result;
    }

    g3(): number {
        const result: number = (this.view.getUint8(this.pos++) << 16) | this.view.getUint16(this.pos);
        this.pos += 2;
        return result;
    }

    g4(): number {
        const result: number = this.view.getInt32(this.pos);
        this.pos += 4;
        return result;
    }

    g8(): bigint {
        const result: bigint = this.view.getBigInt64(this.pos);
        this.pos += 8;
        return result;
    }

    gsmart(): number {
        return this.view.getUint8(this.pos) < 0x80 ? this.g1() - 0x40 : this.g2() - 0xc000;
    }

    // signed
    gsmarts(): number {
        return this.view.getUint8(this.pos) < 0x80 ? this.g1() : this.g2() - 0x8000;
    }

    gjstr(): string {
        const start: number = this.pos;
        const view: DataView = this.view;
        const length: number = view.byteLength;
        while (this.pos < length && view.getUint8(this.pos) !== 10) {
            this.pos++;
        }
        const bytes: Uint8Array = this.data.subarray(start, this.pos);
        if (this.pos < length) {
            this.pos++; // skip terminator
        }
        return new TextDecoder().decode(bytes);
    }

    gdata(length: number, offset: number, dest: Uint8Array | Int8Array): void {
        dest.set(this.data.subarray(this.pos, this.pos + length), offset);
        this.pos += length;
    }

    p1isaac(opcode: number): void {
        this.view.setUint8(this.pos++, (opcode + (this.random?.nextInt ?? 0)) & 0xff);
    }

    p1(value: number): void {
        this.view.setUint8(this.pos++, value);
    }

    p2(value: number): void {
        this.view.setUint16(this.pos, value);
        this.pos += 2;
    }

    ip2(value: number): void {
        this.view.setUint16(this.pos, value, true);
        this.pos += 2;
    }

    p3(value: number): void {
        this.view.setUint8(this.pos++, value >> 16);
        this.view.setUint16(this.pos, value);
        this.pos += 2;
    }

    p4(value: number): void {
        this.view.setInt32(this.pos, value);
        this.pos += 4;
    }

    ip4(value: number): void {
        this.view.setInt32(this.pos, value, true);
        this.pos += 4;
    }

    p8(value: bigint): void {
        this.view.setBigInt64(this.pos, value);
        this.pos += 8;
    }

    pjstr(str: string): void {
        const encoded: Uint8Array = new TextEncoder().encode(str);
        this.data.set(encoded, this.pos);
        this.pos += encoded.length;
        this.view.setUint8(this.pos++, 10);
    }

    pdata(src: Uint8Array, length: number, offset: number): void {
        this.data.set(src.subarray(offset, offset + length), this.pos);
        this.pos += length - offset;
    }

    psize1(size: number): void {
        this.view.setUint8(this.pos - size - 1, size);
    }

    bits(): void {
        this.bitPos = this.pos << 3;
    }

    bytes(): void {
        this.pos = (this.bitPos + 7) >>> 3;
    }

    gBit(n: number): number {
        let bytePos: number = this.bitPos >>> 3;
        let remaining: number = 8 - (this.bitPos & 7);
        let value: number = 0;
        this.bitPos += n;

        for (; n > remaining; remaining = 8) {
            value += (this.view.getUint8(bytePos++) & Packet.bitmask[remaining]) << (n - remaining);
            n -= remaining;
        }

        if (n === remaining) {
            value += this.view.getUint8(bytePos) & Packet.bitmask[remaining];
        } else {
            value += (this.view.getUint8(bytePos) >>> (remaining - n)) & Packet.bitmask[n];
        }

        return value;
    }

    rsaenc(mod: bigint, exp: bigint): void {
        const length: number = this.pos;
        this.pos = 0;

        const temp: Uint8Array = new Uint8Array(length);
        this.gdata(length, 0, temp);

        const bigRaw: bigint = bytesToBigInt(temp);
        const bigEnc: bigint = bigIntModPow(bigRaw, exp, mod);
        const rawEnc: Uint8Array = bigIntToBytes(bigEnc);

        this.pos = 0;
        this.p1(rawEnc.length);
        this.pdata(rawEnc, rawEnc.length, 0);
    }
}
